rfc/rfc-0005-customizable-consumption-offset.md

```diff
@@ -0,0 +1,49 @@
+# RFC-#: Title
+
+**Authors:** Eric Bottard
+
+**Status:**
+
+**Pull Request URL:** link
+
+**Superseded by:** N/A
+
+**Supersedes:** N/A
+
+**Related:** rfc-0003: developer utils
+
+
+## Problem
+Sometimes users want to replay data through functions, maybe because the function logic has changed, or maybe because downstream systems have lost state and that state needs to be re-created. The current Processor concept in riff doesn't support such a mechanism though, and always consumes data from the (current) end of streams.
+
+### Anti-Goals
+
+## Solution
+With a top to bottom view of the user experience, the proposed solution is to
+- introduce an offset option on each input binding when a processor is created
+- persist that option in the Processor CRD instance on the cluster
+- have the Processor reconciler forward that option to the streaming-processor, so that it changes its SubscriptionRequests accordingly
+
+### User Impact
+This RFC proposes to change the `riff streaming processor create` command from
+```
+riff streaming processor create <name> ... --input [alias:]stream ...
+```
+to
+```
+riff streaming processor create <name> ... --input [alias:]stream[:offset] ...
+```
+where `[..]` denotes optional values and `[:offset]` could take values `start` or `end`. Those two logical values are the only ones currently supported by liiklus (and the high level kafka consumer as far as the author can remember). Nevertheless, the encoding of those two values could be done via negative numerical values, in preparation of future support of an actual numerical index.
+
+
+The actual behavior of this proposed change has to be well understood: the offset parameter only makes sense for a _new consumer group_, _i.e._ a group which has never been used and for which liiklus doesn't currently have positions. As soon as the group starts to be used and its positions stored, the `start` value will have no effect. Indeed, this is desired behavior otherwise everytime a processor scales down and up again, it would replay everything over and over again. The corollary from this is that to start consuming from the beginning *again*, a processor would need to be *re-created*, with a different name (because the consumer group is based on the name of the processor).
```

Comment:
how are the consumer groups managed when there are multiple consumers? What connects a specific workload to a specific consumer group at runtime as processors scale up and down?

Reply:
currently there's a 1:1 relationship between a processor resource and a consumer group, but we should consider an intermediate resource to enable a 1:n relationship so that a processor could be replayed (or eventually played from an explicit offset)... for example in a case where the function is updated to fix a bug, and the entire event log should be re-processed

Reply:
I'm not sure I understand the rationale. A processor is just a way to stitch together a function and some streams. If you want another run (with a different group) of that combination, just create another processor with the same combination.

Also, while I find those questions interesting (and took some time to dig into actual liiklus code), I don't think this should happen here. This (light) rfc was opened to get quick consensus on the way the start offset feature should "feel".

Reply:
One of the values of an RFC is that it can stimulate ideas that we might not have considered otherwise, and I think the notion that a Processor may have identity that spans replay is worth considering (similar to how a Deployment has identity that spans sequences of ReplicaSets). If we decide there's merit to that approach, it would mean potentially not adding the one-off configuration option for this at the Processor level. Typically a Processor will have a long lifespan (continuing to read from the "latest" end of the stream until deleted), so I think we should question whether the offset as a configuration option that is set only at initialization time is the right approach.

Reply:
> (because the consumer group is based on the name of the processor)

For the same reasons that stream names should not translate directly to topics, processor names should not translate to consumer groups.

Reply:
> For the same reasons that stream names should not translate directly to topics, processor names should not translate to consumer groups.

I take those reasons are inclusion of the namespace and collision avoidance. Still, in the same way that there is a 1:1 mapping between streams and topics, there is currently a 1:1 mapping between processor and consumer group (also, created [this](https://github.com/projectriff/system/issues/226)). 

I'm fine considering departure from that model, in this RFC or another one if we want to move forward in a timely fashion. I'm not sure I understand what you mean @markfisher by "identity that spans replay" though. Currently, the processor identity captures a function + a set of streams. Tomorrow it may also encompass the willingness to "start from the beginning for stream X, Y, Z". Are you saying _that whole set of characteristics_ is "the identity" and one may want to "do it again", ie "(re-)create a «processor» that uses the same set of start-offset semantics"? If so, how is that not creating a processor with a new name? Alternatively, should we add something like `riff streaming processor <foo> replay`?

Reply:
If we depart from the model, it can definitely be in a different RFC and a longer term initiative. I just want to make sure we don't introduce something that makes that more difficult later. Perhaps a way to consider my suggestion that seems less extreme - and more aligned with the way you just described it @ericbottard - would be to consider the new layer of resource being added _above_ the current Processor resource instead of _beneath_.